File: dataStructures/infixPostfixConvert.py
import stack

def infixToPostfix(expr):
    s = stack.Stack()
    prec = {"*":3, "/":3, "+":2, "-":2, "(":1}
    ops = "*/+-("
    postfixString = ""
    for char in expr:
        if (char in ops):
            lower = False
            while (not s.isEmpty() and not lower and char != '('):
                op = s.peek()
                if (prec[op] < prec[char]):
                    lower = True
                else:
                    postfixString += s.pop()
            s.push(char)
        elif (char == ')'):
            op = s.pop()
            while (op != '('):
                postfixString += op
                op = s.pop()
        else:
            postfixString += char
    while (not s.isEmpty()):
        postfixString += s.pop()
    return postfixString


#print infixToPostfix("A+B*C")
#print infixToPostfix("A+B+C+D")
#print infixToPostfix("A+B*C+D")
print infixToPostfix("(A+B)*(C+D)")
#print infixToPostfix("A*B+C*D")

File: dataStructures/infixPostfixConvert.rb
require_relative 'stack'

def infix_postfix_convert(expr)
  prec = {"*" => 3, "/" => 3, "+" => 2, "-" => 2, "(" => 1}
  s = Stack.new
  operands = "+-/*()"
  str = ""
  expr.each_char do |c|
    if (operands.include?(c))
      if (c == '(')
        s.push(c)
      elsif (c == ')')
        op = s.pop
        while (op != '(')
          str += op
          op = s.pop
        end
      else
        while (!s.is_empty? and prec[c] < prec[s.peek])
          str += s.pop
        end
        s.push(c)
      end
    else
      str += c
    end
  end
  while (!s.is_empty?)
    str += s.pop
  end
  return str
end


begin
  inf = "2*3/4"
  puts infix_postfix_convert(inf)
end
